Dalamud/Plugin/Services/IHookProvider.cs

```diff
@@ -0,0 +1,97 @@
+﻿using System.Diagnostics;
+
+using Dalamud.Hooking;
+using Dalamud.Utility.Signatures;
+
+namespace Dalamud.Plugin.Services;
+
+/// <summary>
+/// Service responsible for the creation of hooks.
+/// </summary>
+public interface IHookProvider
+{
+    /// <summary>
+    /// Available hooking backends.
+    /// </summary>
+    public enum HookBackend
+    {
+        /// <summary>
+        /// Choose the best backend automatically.
+        /// </summary>
+        Automatic,
+        
+        /// <summary>
+        /// Use Reloaded hooks.
+        /// </summary>
+        Reloaded,
+        
+        /// <summary>
+        /// Use MinHook.
+        /// You should never have to use this without talking to us first.
+        /// </summary>
+        MinHook,
+    }
+    
+    /// <summary>
+    /// Initialize <see cref="Hook{T}"/> members decorated with the <see cref="SignatureAttribute"/>.
+    /// Errors for fallible signatures will be logged.
+    /// </summary>
+    /// <param name="self">The object to initialise.</param>
+    public void InitializeFromAttributes(object self);
```

Comment:
Will this method also be responsible for initializing and setting plain delegates? If so, consider documenting this as doing more than just hooks.

I'm also not particularly a fan of delegate management being handled by `IHookProvider` semantically, but I suppose it makes enough sense (and likely isn't worth the effort to make into a separate system).

Reply:
I believe it should absolutely only handle hooks and not the other stuff `SignatureHelper` does. It would just result in another inconsistent API, which will require another refactor in the future.

For delegates, we could introduce a new `MemoryHelper` method:

```cs
public static T GetDelegateForSignature<T>(string signature) where T : Delegate
{
    var scanner = Service<SigScanner>.Get();
    var address = scanner.ScanText(signature);
    return Marshal.GetDelegateForFunctionPointer<T>(address);
}
```

Then, plugins can use it like this:

```cs
using Dalamud.Memory;

public class GameFunctions
{
    private delegate byte IsQuestCompletedDelegate(ushort questId);
    private readonly IsQuestCompletedDelegate IsQuestCompleted = MemoryHelper.GetDelegateForSignature<IsQuestCompletedDelegate>("E8 ?? ?? ?? ?? 41 88 84 2C");
}
```

It's an easy to understand alternative.  
Though the line might get a little long, it doesn't need an attribute and there is no need to call `SignatureHelper.Initialise(this)` the constructor.
It also throws, just like `Hook<T>.FromSignature()`, if the signature is not found, making it undeniably clear that the signature is outdated.

Besides delegates, `SignatureHelper` can also fill a fields **initial** value. But do we *really* need to support this? It's easy enough to read it in the constructor, to create a property with a getter, or to use Lazy:

```cs
public Lazy<nint> SomeAddress => new(() => MemoryHelper.Read<nint>(Service.SigScanner.GetStaticAddressFromSig(sig)));
```

🤷‍♂️

Reply:
I generally agree, but I think that crippling an API on purpose for correctness' sake is a little unfortunate. Not 100% sure what to do here. Looking through codesearch, hooks, delegates and address modes are used quite frequently in plugins. I would honestly tend to just leaving it like this.

Maybe the service should be renamed to IGameInteropProvider, and docs should be improved, so that it's clearer that it's not just hooks?

Reply:
Docs plus a better name would be nice. Maybe `INativeInteropService` or similar? These methods can be used for more than just game hooks, after all.

Reply:
I think `INativeInteropService` is fine. It just sounds a bit like it should be the home of the functions in `MemoryHelper`. Not sure if that's a good thing. 🤔

Anyhow, if you use a broader name, I’d suggest that the method names should reflect their usage then, i.e. renaming `FromSignature` to `HookFromSignature` and so on, so that they don't conflict with other things that might be created from a signature in the future, like `DelegateFromSignature`. 👀

Reply:
I addressed both of this, it's a bit more verbose now, but it does feel better overall IMO.

Reply:
@Haselnussbomber w/r/t `DelegateFromSignature`: one can already do that from SigScanner pretty reliably. We can add to this later, but it's not like we're removing any functionality; I think we're good.